Adding an app for redirects when storing state in session storage

kobelb opened the pull request:
When using the `state:storeInSessionStorage` setting with the short-urls, we were previously doing a `reply().redirect(fullUrl)` and then the state was being extracted from the URL and being put into sessionStorage.  This was causing issues with IE where the URL was too long.

This approach uses a separate 'app' and a variable that it's grabbing from the kbn-initial-state to get all the information to the browser, where it then puts it in sessionStorage before redirecting the user to the URL with the shortened hashes. It's not an incredibly elegant solution, and the user experience is somewhat 'choppy' for lack of a better description.

I'm open to alternatives to the current approach, but what's in this PR should work.

Resolves #10128

kobelb commented:
jenkins test this

spalger commented on src/server/http/index.js:
```diff
@@ -120,8 +120,19 @@ module.exports = async function (kbnServer, server, config) {
+        const uiSettings = server.uiSettings();
+        const stateStoreInSessionStorage = await uiSettings.get(request, 'state:storeInSessionStorage');
+        if (!stateStoreInSessionStorage) {
+          reply().redirect(config.get('server.basePath') + url);
```
Make sure to return here.

spalger replied:
Make sure to redirect here

spalger commented on src/core_plugins/state_session_storage_redirect/public/index.js:
```diff
@@ -0,0 +1,77 @@
+};
+
+import uiRoutes from 'ui/routes';
+uiRoutes.enable();
```
Is routing really required? Adding the router causes all of the routeSetupWork to execute, which includes several API calls that will make this take longer than necessary.

spalger replied:
Do we really need the router here? Enabling the router runs all of the route setup work, which just slows down this simple page.

kobelb replied:
Makes sense, I'll remove it.

spalger commented on src/core_plugins/state_session_storage_redirect/public/index.js:
```diff
@@ -0,0 +1,77 @@
+import { parse as parseUrl, format as formatUrl } from 'url';
+import { stringify as stringifyQuerystring } from 'querystring';
+
+const conservativeStringifyQuerystring = (query) => {
```
I think this logic should be put in the `state_management` module next to src/ui/public/state_management/state_hashing/unhash_url.js, and it definitely needs tests.

spalger replied:
All of this hashing logic should be put in the `state_management` module next to src/ui/public/state_management/state_hashing/unhash_url.js, and it definitely needs tests.

spalger commented on src/server/http/index.js:
```diff
@@ -120,8 +120,19 @@ module.exports = async function (kbnServer, server, config) {
+          redirectUrl: url,
+        });
       } catch (err) {
+        console.log(err);
```
Make sure to remove this

kobelb replied:
Good catch, thanks!

spalger commented on src/ui/index.js:
```diff
@@ -83,12 +83,12 @@ export default async (kbnServer, server, config) => {
         uiExports.injectedVarsReplacers,
         async (acc, replacer) => await replacer(acc, request, server),
-        defaults(await app.getInjectedVars() || {}, uiExports.defaultInjectedVars)
+        defaults(await app.getInjectedVars() || {}, uiExports.defaultInjectedVars, vars)
```
I think we should rename `vars` -> `injectedVarsOverrides` and move it to the front of the list here. If the code calling `renderApp()` has specific vars in mind they shouldn't be overridden by the defaults or the standard injected vars, only the "replacers"  could override them.

kobelb commented:
@spalger I've addressed your concerns/comments, except the comment about not using the routing. I tried to do the following to remove the need for routing https://github.com/elastic/kibana/pull/10822/commits/14b09e3282803737b540fb46e4094440984d3172 but this led to the code being run whenever the javascript was loaded, is there a different way that I should be removing the need to use routing?

spalger commented:
>  this led to the code being run whenever the javascript was loaded

I'm not sure I understand the issue.

kobelb commented:
If the following is run every-time that load Kibana in a browser:

```
const redirectUrl = chrome.getInjected('redirectUrl');

 const hashedUrl = hashUrl([new AppState(), globalState], redirectUrl);
const url = chrome.addBasePath(hashedUrl);

$window.location = url;
```

it'll always try to redirect the user, when we only want it to be ran after the following call:

```
 const app = kbnServer.uiExports.apps.byId.stateSessionStorageRedirect;
 reply.renderApp(app, {
  redirectUrl: url,
});
```

Stacey-Gammon commented:
fixes https://github.com/elastic/kibana/issues/10128

kobelb commented on src/ui/public/state_management/state_hashing/hash_url.js:
```diff
@@ -0,0 +1,77 @@
+  });
+};
+
+export default hashUrl;
```
you got it! old habits die hard...

kobelb commented on src/core_plugins/state_session_storage_redirect/package.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "state_session_storage_redirect",
+  "version": "kibana"
```
Good call, I'll add a description in here.

kobelb commented:
Backported

5.x - https://github.com/elastic/kibana/commit/9ca5059c467185b298263647d47fb5d43fbfa339
